ops/ops/interface_aws/requires.py

```diff
@@ -0,0 +1,289 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""Implementation of aws interface.
+
+This only implements the requires side, currently, since the providers
+is still using the Reactive Charm framework self.
+"""
+import json
+from hashlib import sha256
+import logging
+import ops
+import string
+from functools import cached_property
+from typing import  Mapping, Optional
+from urllib.parse import urljoin
+from urllib.request import urlopen, Request
+
+
+log = logging.getLogger(__name__)
+
+# block size to read data from AWS metadata service
+# (realistically, just needs to be bigger than ~20 chars)
+READ_BLOCK_SIZE = 2048
+
+# the IP is the AWS metadata service, documented here:
+# https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html
+_METADATAV2_TOKEN_URL = "http://169.254.169.254/latest/api/token"
+_METADATA_URL = "http://169.254.169.254/latest/meta-data/"
+_INSTANCE_ID_URL = urljoin(_METADATA_URL, "instance-id")
+_AZ_URL = urljoin(_METADATA_URL, "placement/availability-zone")
+
+def _imdv2_request(url):
+    token_req = Request(
+        _METADATAV2_TOKEN_URL,
+        headers={"X-aws-ec2-metadata-token-ttl-seconds": "21600"},
+    )
+    setattr(token_req, "method", "PUT")
+
+    with urlopen(token_req) as fd:
+        token = fd.read(READ_BLOCK_SIZE).decode("utf8")
+        return Request(url, headers={"X-aws-ec2-metadata-token": token})
```

Comment:
Might we want to raise if the http request fails and then display a status in Juju?

Reply:
yeah, i suppose.  lemme see if there's a fair way to do something about that.